File: infra/aos-assembly/Aos.ControlPlane/Components/Events/EventNotifications.cs
using System.Collections.Generic;
using Aos.DataModel;
using Aos.PlanarFoundation;
using Pulumi;
using Pulumi.AwsNative.Events;
using Pulumi.AwsNative.Events.Inputs;
using Pulumi.AwsNative.Iam;
using Pulumi.AwsNative.Logs;
using Pulumi.AwsNative.Sqs;

namespace Aos.ControlPlane.Components.Events;

public class EventNotifications : ComponentResource
{
    public LogGroup EventStreamLogs { get; set; }
    public Queue EventDedupeQueue { get; set; }
    public EventBus Bus { get; set; }
    public EventForwarder Forwarder { get; set; }
    
    public EventNotifications(ClusterSpec spec, string name, ComponentResourceOptions opts)
        : base("autoos:controlplane:EventNotifications", name, opts)
    {
        this.EventDedupeQueue = new Queue("DedupeQueue", new QueueArgs()
        {
            QueueName = spec.GetHyphenatedPrefix() + "-maestro-deduper.fifo",
            ContentBasedDeduplication = true,
            FifoQueue = true,
        });
        this.Bus = new EventBus("Bus", new EventBusArgs()
        {
            EventSourceName = spec.GetHyphenatedPrefix() + "-maestro"
        });
        this.EventStreamLogs = new LogGroup("StreamedEvents", new LogGroupArgs()
        {
            LogGroupName = spec.GetTitledPrefix() + "MaestroEvents"
        });
        this.EventSubmittedPolicyArn =
            Output.All<string>(Bus.Arn, EventDedupeQueue.Arn).Apply(arr =>
            {
                var busArn = arr[0];
                var queueArn = arr[1];
                var mp = new ManagedPolicy("SubmitterPolicy", new ManagedPolicyArgs()
                {
                    ManagedPolicyName = spec.GetTitledPrefix() + "MaestroEventSubmitter",
                    PolicyDocument = IamExtensions.SimpleDoc(
                        new IamExtensions.SimpleStatement(new []{ "events:PutEvents" }, new []{ busArn }),
                        new IamExtensions.SimpleStatement(new []{ "sqs:SendMessage", "sqs:SendMessageBatch" }, new []{ queueArn }))
                }, new CustomResourceOptions() { Parent = this });
                return mp.PolicyArn;
            });

        new Rule("BakeryRule", new RuleArgs()
        {
            Name = spec.GetTitledPrefix() + "MaestroBakeryRule",
            EventPattern = new Dictionary<string, object?>
            {
                ["detailType"] = new[] { "EC2 Image Builder Image State Change" },
                ["source"] = new[] { "aws.imagebuilder" },
                ["resources"] = new[]
                {
                    new Dictionary<string, object?>()
                    {
                        { "prefix", "2017-10-02" }
                    }
                }
            }
        }, new CustomResourceOptions() { Parent = this });
        
        new Rule("BroadcastRule", new RuleArgs()
        {
            Name = spec.GetTitledPrefix() + "MaestroBroadcastRule",
            EventBusName = this.Bus.Name,
            EventPattern = new Dictionary<string, object?>
            {
                ["source"] = new[] { "maestro" }
            },
            Targets = new InputList<RuleTargetArgs>()
            {
                new RuleTargetArgs()
                {
                    Arn = this.EventStreamLogs.Arn
                }
            }
        }, new CustomResourceOptions() { Parent = this });

        this.Forwarder = new EventForwarder(new EventForwarderParams()
        {
            EventBusName = this.Bus.Name,
            DedupeQueueArn = this.EventDedupeQueue.Arn
        },"Forwarder", new ComponentResourceOptions() { Parent = this });
    }
    
    public Output<string> EventSubmittedPolicyArn { get; set; }
}

File: infra-constructs/aos-cluster-constructs/lib/controlPlane/EventNotifications.ts
import { aws_events as evts, aws_events_targets as targets, aws_iam as iam, aws_logs as logs, aws_sqs as sqs } from "aws-cdk-lib";
import { Construct } from "constructs";
import * as cdk from 'aws-cdk-lib';
import * as path from "path";
import * as fs from "fs";
import { EventForwarder } from "./EventForwarder";
import { fromParams } from "../common";

const code = fs
  .readFileSync(path.join(__dirname, "./cleanup.py"))
  .toString("utf8");

export class EventNotifications extends Construct {
  readonly eventStreamLogs: logs.LogGroup;
  readonly eventDedupeQueue: sqs.Queue;
  readonly bus: evts.EventBus;
  readonly eventSubmitterPolicy: iam.ManagedPolicy;
  readonly forwarder: EventForwarder;

  constructor(parent: Construct, name: string) {
    super(parent, name);

    const info = fromParams(this);
    const names = info.namesFor(this);

    this.eventDedupeQueue = new sqs.Queue(this, "EventDedupeQueue", {
      queueName: names.hyphenatedLowercasePrefix + `maestro-deduper.fifo`,
      contentBasedDeduplication: true,
      fifo: true,
    });
    this.bus = new evts.EventBus(this, "Events", {
      eventBusName: names.hyphenatedLowercasePrefix + `maestro`,
    });
    this.eventStreamLogs = new logs.LogGroup(this, "EventStreamLogs", {
      logGroupName: names.hyphenatedPrefix + `MaestroEvents`,
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    this.eventSubmitterPolicy = new iam.ManagedPolicy(
      this,
      "EventSubmitterPolicy",
      {
        managedPolicyName:
          names.hyphenatedLowercasePrefix + `MaestroEventSubmitter`,
        statements: [
          new iam.PolicyStatement({
            actions: ["events:PutEvents"],
            resources: [this.bus.eventBusArn],
          }),
          new iam.PolicyStatement({
            actions: ["sqs:SendMessage", "sqs:SendMessageBatch"],
            resources: [this.eventDedupeQueue.queueArn],
          }),
        ],
      }
    );
    
    const amiBakeryRule = new evts.Rule(this, "BakeryRule", {
      ruleName: names.hyphenatedLowercasePrefix + `MaestroBakeryRule`,
      eventPattern: {
        detailType: ["EC2 Image Builder Image State Change"],
        source: ["aws.imagebuilder"],
        resources: evts.Match.prefix(
          `arn:aws:imagebuilder:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:image/opensearch`
        ),
      },
    });

    const broadcastRule = new evts.Rule(this, "MaestroEventBroadcaster", {
      ruleName: names.hyphenatedLowercasePrefix + `MaestroBroadcastRule`,
      eventBus: this.bus,
      eventPattern: {
        source: ["maestro"],
      },
    });
    broadcastRule.addTarget(
      new targets.CloudWatchLogGroup(this.eventStreamLogs)
    );
    this.forwarder = new EventForwarder(this, "Forwarder", {
      eventBus: this.bus,
      dedupeQueue: this.eventDedupeQueue,
    });
    amiBakeryRule.addTarget(new targets.LambdaFunction(this.forwarder.func, {}));
  }
}
